base/broadcast.jl

```diff
@@ -612,63 +776,183 @@ julia> string.(("one","two","three","four"), ": ", 1:4)
-    if isempty(iter)
-        return Base.similar(Array{ElType}, shape)
+## general `copyto!` methods
+# The most general method falls back to a method that replaces Style->Nothing
+# This permits specialization on typeof(dest) without introducing ambiguities
+@inline copyto!(dest::AbstractArray, bc::Broadcasted) =
+    copyto!(dest, convert(Broadcasted{Nothing}, bc))
+
+# Performance optimization for the Scalar case
+@inline function copyto!(dest::AbstractArray, bc::Broadcasted{<:Union{Scalar,Unknown},ElType,Nothing,Nothing}) where ElType
+    if not_nested(bc)
+        if bc.f === identity && bc.args isa TupleLL{<:Any,TupleLLEnd} # only a single input argument to broadcast!
+            # broadcast!(identity, dest, val) is equivalent to fill!(dest, val)
+            return fill!(dest, bc.args.head)
+        else
+            args = Tuple(bc.args)
+            @inbounds for I in eachindex(dest)
+                dest[I] = bc.f(args...)
+            end
+            return dest
+        end
     end
-    keeps, Idefaults = map_newindexer(shape, As)
-    st = start(iter)
-    I, st = next(iter, st)
-    val = f([ _broadcast_getindex(As[i], newindex(I, keeps[i], Idefaults[i])) for i=1:nargs ]...)
-    if val isa Bool
-        dest = Base.similar(BitArray, shape)
-    else
-        dest = Base.similar(Array{typeof(val)}, shape)
+    # Fall back to the default implementation
+    copyto!(dest, instantiate(instantiate_axes(bc)))
+end
+
+# Specialize this method if all you want to do is specialize on typeof(dest)
+@inline function copyto!(dest::AbstractArray, bc::Broadcasted{Nothing})
+    axes(dest) == axes(bc) || throwdm(axes(dest), axes(bc))
+    # Performance optimization: broadcast!(identity, dest, A) is equivalent to copyto!(dest, A) if indices match
+    if bc.f === identity && bc.args isa TupleLL{<:AbstractArray,TupleLLEnd} # only a single input argument to broadcast!
+        A = bc.args.head
+        if axes(dest) == axes(A)
+            return copyto!(dest, A)
+        end
     end
-    dest[I] = val
-    _broadcast!(f, dest, keeps, Idefaults, As, Val(nargs), iter, st, 1)
-end
-
-broadcast(f, ::Union{Scalar,Unknown}, ::Nothing, ::Nothing, a...) = f(a...)
-
-@inline broadcast(f, ::Style{Tuple}, ::Nothing, ::Nothing, A, Bs...) =
-    tuplebroadcast(f, longest_tuple(A, Bs...), A, Bs...)
-@inline tuplebroadcast(f, ::NTuple{N,Any}, As...) where {N} =
-    ntuple(k -> f(tuplebroadcast_getargs(As, k)...), Val(N))
-@inline tuplebroadcast(f, ::NTuple{N,Any}, ::Type{T}, As...) where {N,T} =
-    ntuple(k -> f(T, tuplebroadcast_getargs(As, k)...), Val(N))
-longest_tuple(A::Tuple, B::Tuple, Bs...) = longest_tuple(_longest_tuple(A, B), Bs...)
-longest_tuple(A, B::Tuple, Bs...) = longest_tuple(B, Bs...)
-longest_tuple(A::Tuple, B, Bs...) = longest_tuple(A, Bs...)
-longest_tuple(A, B, Bs...) = longest_tuple(Bs...)
-longest_tuple(A::Tuple) = A
+    @simd for I in CartesianIndices(axes(bc))
+        @inbounds dest[I] = _broadcast_getindex(bc, I)
```

Comment:
Nifty that you can index directly into the `Broadcasted` objects.  Have you thought about making this a bit more first class?  Like calling it simply `getindex(::Broadcasted, …)`?  Or maybe even `Broadcasted <: AbstractArray`?  Not crucial by _any_ means.  Just cool.

Reply:
Yes, I have, and I've gone back and forth on that. Currently I haven't done that for one somewhat arcane reason: in `A .+ (v .+ 1)` (`A` is a matrix and `v` a vector), the inner broadcasting operation with `v` doesn't have any objects that have 2d indices. Consequently it seems that `bcinner[i, j]` should throw a `BoundsError` if `j>1`, but `_broadcast_getindex` doesn't fall victim to that requirement.

Now, you might argue that because `bcinner` will typically be instantiated, the `axes` field indicates that really, it *does* have suitable indices. But not all `Broadcasted` types get instantiated (e.g., for StaticArrays one may want to short-circuit the `axes` field computation). Since in such cases there's no hint that `bcinner` should be supported over a larger domain, I decided not to call it `getindex`.

Like I said, pretty arcane.

Reply:
Could we define `getindex` just on `Broadcasted` objects that have instantiated indexers?  The main reason to consider this is that it provides a very nice public & stable API for working with `Broadcasted` objects.  Folks will need to handle them in their `copyto!` methods, but it'd be nice to be able to draw some clear lines within Base.Broadcast on what's public and what's private.

How are you imagining StaticArrays would use this?  `Broadcast.flatten`?